### PR TITLE
[primTorch] Add refs for data conversion ops

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #86561

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Related issue: https://github.com/pytorch/pytorch/issues/86558

**Comment by nkaretnikov** [comment]
Note: the builtins changes are due to name clashes with added refs.

**Review by lezcano — COMMENTED** [review]
The refs (and the issue) LGTM. The only point is that I think that you have missed `chalf`, `cdouble` and `cfloat`.

I'm concerned about having to use `builtins.int` and so on throughout the whole file. There must be a better way? @mruberry may be able to come up with something.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -14159,6 +14159,32 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
             # RuntimeError: attribute lookup is not defined on builtin
             DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
         )),
+    UnaryUfuncInfo(
+        'cdouble',
+        op=lambda x, *args, **kwargs: x.cdouble(*args, **kwargs),
```
Comment (lezcano) [inline torch/testing/_internal/common_methods_invocations.py:14164]: Can we just write `op=Tensor.cdouble`? I think the answer is no, but I don't remember the reason off the top of my head.

**Comment by mruberry** [comment]
> I'm concerned about having to use `builtins.int` and so on throughout the whole file. There must be a better way? @mruberry may be able to come up with something.

I agree. First, great diligence being so thorough @nkaretnikov. But what if we put these references into a `_conversions` submodule with a note that it's because we don't want to clobber the builtin names elsewhere? When we map from the torch operation to its reference we can have the mapping go to the private `_conversions` submodule without issue. This may be a little confusing if someone is trying to use the refs directly (because this would be the only case where the `refs` namespace doesn't correspond with the `torch` namespace structure), but we can fix that later if it turns out to be an important scenario for these operations.

What do you think of that approach, @lezcano?

**Comment by nkaretnikov** [comment]
@lezcano @mruberry PTAL.

**Review by lezcano — APPROVED** [review]
Cool!

File: test/functorch/test_ops.py [hunk 2]
```diff
@@ -589,6 +589,8 @@ def fn(inp, *args, **kwargs):
         xfail("as_strided_scatter"),  # incorrect output
         skip("bernoulli"),  # calls random op
         xfail("bfloat16"),  # rank 4 tensor for channels_last
+        xfail("cdouble"),  # rank 4 tensor for channels_last
```
Comment (mruberry) [inline test/functorch/test_ops.py:592]: fyi @zou3519

File: torch/_refs/_conversions.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import torch
+
+from torch._prims_common import TensorLikeType
+
+# Data conversion references.
+#
+# Note: this module breaks the usual _refs to torch naming scheme where
```
Comment (mruberry) [inline torch/_refs/_conversions.py:7]: Nice comment

File: torch/_refs/_conversions.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+# _refs.foo.bar is a ref for torch.foo.bar.  The following definitions are not
+# part of _refs/__init__.py to avoid name clashes with Python builtin types
+# (like int).
+
+__all__ = [
+    "bfloat16",
```
Comment (mruberry) [inline torch/_refs/_conversions.py:13]: Nit: alphabetical order, please

File: torch/_refs/_conversions.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    "int",
+    "long",
+    "short",
+    "chalf",
+    "cfloat",
+    "cdouble",
+]
+
+
+def bool(
```
Comment (mruberry) [inline torch/_refs/_conversions.py:29]: Create a factory to generate the conversions:

```
def _make_conversion_method(name, dtype):
  def fn(self, memory_format):
    return self.to(dtype, memory_format=memory_format)
  
  fn.__name__ = name
  return fn
```

**Review by mruberry — APPROVED** [review]
Looks good, but let's give @zou3519 a chance to look. Also see my suggestion to create a factory for the conversion methods. That should ensure more uniformity for them.

**Review by zou3519 — APPROVED** [review]
lgtm from functorch side, all the xfails are pre-existing conditions

**Pull request closed without merging**